src/lib/user.ts

```diff
@@ -0,0 +1,79 @@
+import gravatarUrl from 'gravatar-url';
+import Joi from 'joi';
+
+export interface UserData {
+    id?: number;
+    isAPI?: boolean;
+    name?: string;
+    username?: string;
+    email?: string;
+    permissions?: string[];
+    imageUrl?: string;
+    seenAt?: Date;
+    loginAttempts?: number;
+    createdAt?: Date;
+}
+
+export default class User {
+    id: number;
+
+    isAPI: boolean;
+
+    name: string;
+
+    username: string;
+
+    email: string;
+
+    permissions: string[];
+
+    imageUrl: string;
+
+    seenAt: Date;
+
+    loginAttempts: number;
+
+    createdAt: Date;
+
+    constructor(
+        {
+            id,
+            isAPI,
+            name,
+            email,
+            username,
+            imageUrl,
+            permissions,
+            seenAt,
+            loginAttempts,
+            createdAt,
+        }: UserData = { isAPI: false },
+    ) {
+        if (!username && !email) {
+            throw new TypeError('Username or Email us required');
+        }
+        Joi.assert(email, Joi.string().email(), 'Email');
+        Joi.assert(username, Joi.string(), 'Username');
+        Joi.assert(name, Joi.string(), 'Name');
+
+        this.id = id;
+        this.isAPI = isAPI;
```

Comment:
@chriswk this feels a bit confusing. Suggestion on how to differentiate a "API user" from a "human user"?

Reply:
Having a boolean column in the user table for whether or not it's an API user doesn't seem that complicated. 
Or maybe have a separate class/type/table for ApiUsers, that complicates join operations though, but strictly speaking they are different concepts?

Reply:
lets keep it until we have done the api tokens thing in #755